README.TXT
==================== Media Server Website ====================
Developed by: Brian Cullinan
Version: 0.4.0 (Alpha)
Release Date: 2009-02-18

Unpack or Checkout to your server directory.
Read /<site path>/include/settings.<os>.php for more information.
Install by browsing to <site path>/admin/index.php.

For Developers see DEVELOPERS.TXT
For changes see CHANGELOG.TXT
For todo list and upcoming developments see TODO.TXT
For a list of required software see REQUIREMENTS.TXT


CHANGELOG.TXT
2011-11-26
Bug fixes
Rearranged some code
Worked on tunegazelle template
In the process of removing language stuff
Added basic authentication module, needs further testing
Expanded authentication API

2011-10-16
Started TuneGazelle template
Remove AdoDB, system now uses MySQLi

2011-01-31
Finished language plugin, still need to replace all the text with the lang call
Big fixes

2011-01-20
Worked hard on language plugin, mostly working, Google API needs more corrections

2011-01-13
Compliance corrections
Jquery search textbox and login textboxes

2010-12-26
Worked on Music Downloader script
Fixed up some session features
    Abstracted to new_session() for users.inc module and cli module to use
Started settings.php.dist-gen script to generate all possible settings and their default values
	Just needs some fancy handlers for different options such as text numbers or enums
Added a files_exist_regexp() that checks for files paths using a regular expression
Fixed all the setting_ entries in configure_() functions

2010-12-23
Got it working on Windows again
	Bug fixes
Changed most of the str_replace(DIRECTORY_SEPARATOR and the like
	Still need to verify the database can't be corrupted
	If get_info_() is only called by add() or another get_info_() function then add() can just use str_replace(), otherwise str_replace needs to be used in every get_info_() function

2010-12-22
Finished error logging to remote servers over SSH2
Made a lot of progress on download script
Lots of bug fixes
Added artists wrapper back in to system

2010-12-21
Worked on download_services API
	Downloaders will be able to tie in to the services, make requests, then analyze those requests later
	Download searches are stored in the database for reference and caching
Added API for specifying a function for retrieving url/title pairs from search results for each service
	Many services will use an RSS feed parser in the future
CLI now uses a static session but works from command line, permissions should be determined by the user running the script
Added a cookie jar file setting for curl, intended for use from command line, but may not be useful
Lenveshtein used in searching and pairing discogs and movie modules now doesn't store results that are too inaccurate
Music downloader pairs search results with discogs albums
	Needs to match search results of different formats
	Needs to be abstracted to support more than just an artist as a search term

2010-12-20
Restructures entire Live and Select theme to be more simple, results look very much the same
Fixed error coloring in CLI module
Started download_services interface for downloading from every service automatically
Fixed huge error in cron script, it was erroring out because the error array was getting too big
	This is not intentional behavior for cron script since it does print a ton of status information
	Errors function only stores errors if the error callback does not return false

2010-12-19
Removed use of Filetype = FOLDER since a file on the system could be .folder
	Changed all the entries in the database
	Modified template to use Filemime = httpd/unix-directory
Changed the use of FILE to use a blank Filetype
Added Demonoid to torrent services
Worked on webdav module

2010-12-07
Improved error API greatly
	Error codes are now completely bitwise including E_MODULE and custom error codes
Fixed files.module to use watched.module when possible
Corrections to code, see repo changes

2010-12-03
Improved many conventions with booleans throughout the site
	Appears to be working fine now, settings need to be tested
Added use_https option in general settings
	This is so logins can change to https
	This also sets a secure flag on the session cookie
Removed generic_validate_boolean() functions from the system

2010-11-30
Improved error API
	Errors are stored in a "request" database, so all requests can be referred to
	This database should periodically be purged, and there needs to be
Found one of the longest processes (~5 seconds) is due to counting files in watched.module to verify folder integrity
Configuring errors shows on general settings page
Using a loop for selecting which files are handled by each module provided a decent speed boost, handles_all() in file.inc
Invoke_menu() now closes the session after output_() function is called
Added a 3rd deep select option that also adds information for "cron_only" modules on demand
Removed "$force" option from all add_ functions, system is smart enough now
Removed add_all() from files.module since it was only used by cron so it is now in cron.module
Discogs downloads and displays remote image, this might get slow and needs improvement
Added request printouts to live theme

2010-11-23
Fixed and colorized CLI
	In the process of finishing tools using CLI to debug
Added agent and agent_simple for getting HTTP_USER_AGENT
Added og_theme for getting the contents of an outputted theme

2010-11-21
Rearranged some code
Fixed the output in the search box, was vulnerable to XSS

2010-11-16
Fixed some bugs in url()
	Links should be inputted into url() as fully qualified urlencoded links
	url() decode the link piece by piece, and reconstructs it with the correct menu item
	url() also urlencodes the link
The register_output_vars('get') url was not being urlencoded, but with the corrections to url() it is
Improved speed a little bit by altering the way templates work	
	Added templates/%template/%tfile menu entry
	Changed all links in live styles to use relative path, which will be inherited from /templates/live/
	Template files load directly and don't go through the system

2010-11-15
Fixed watched.module, was scanning every run because handles_watched() wasn't using is_watched() AND handles_files

modules/bt.php
<?php
// handle bit torrenting of selected files

/**
 * Implementation of register
 * @ingroup register
 */
function register_bt()
{
	return array(
		'name' => lang('bt title', 'BitTorrent Creator'),
		'description' => lang('bt description', 'Bittorrent module that creates torrents for files or directories so users can share downloads with each other.'),
		'privilage' => 1,
		'path' => __FILE__,
		'template' => false,
		'depends on' => array('bttracker_installed', 'bttracker_database_installed', 'seeder'),
		'settings' => array('seeder_path', 'seeder_args'),
	);
}

function rewrite_bt($request)
{
	// save the whole request to be used later
	$request['bt_request'] = $request;
}

/**
 * Implementation of setting
 * @ingroup setting
 * @return The default install path for VLC on windows or linux based on validate_SYSTEM_TYPE
 */
function setting_seeder_path($settings)
{
	if(isset($settings['seeder_path']) && is_file($settings['seeder_path']))
		return $settings['seeder_path'];
	else
	{
		if(setting_system_type($settings) == 'win')
			return '';
		else
			return '/usr/local/bin/ctorrent';
	}
}

/**
 * Implementation of setting
 * @ingroup setting
 * @return The entire arg string for further validation by the configure() function
 */
function setting_seeder_args($settings)
{
	if(isset($settings['seeder_args']) && is_file($settings['seeder_args']))
		return $settings['seeder_args'];
	else
	{
		if(setting_system_type($settings) == 'win')
			return '';
		else
			return '-s / "%IF"';
	}
}

/**
 * Implementation of dependency
 * @ingroup dependency
 */
function dependency_seeder($settings)
{
	$settings['seeder_path'] = setting('seeder_path');
	return file_exists($settings['seeder_path']);
}

/**
 * Implementation of dependency
 * @ingroup dependency
 */
function dependency_bttracker_installed($settings)
{
	return (include_path('bttracker/config.php') !== false);
}

/**
 * Implementation of validate
 * @ingroup validate
 * @return announce if 'bt' is set in the request, otherwise returns 'bencode'
 */
function validate_bt($request)
{
	if(isset($request['bt']))
		return 'announce';
	else
		return 'bencode';
}

/**
 * Implementation of validate
 * @ingroup validate
 * @return The entire request for use with the bttracker script
 */
function validate_bt_request($request)
{
	return $request['bt_request'];
}

/**
 * implementation of status
 * @ingroup status
 */
function status_bt($settings)
{
	$status = array();

	if(dependency('bttracker_installed'))
	{
		$status['bttracker_installed'] = array(
			'name' => 'PHP BTTracker+ Installed',
			'status' => '',
			'description' => array(
				'list' => array(
					'The system has detected that bttracker is installed.',
					'Bttracker is an open source torrent tracker.',
				),
			),
			'type' => 'label',
			'value' => 'Bttracker detected',
		);
	}
	else
	{
		$status['bttracker_installed'] = array(
			'name' => 'PHP BTTracker+ Missing',
			'status' => 'fail',
			'description' => array(
				'list' => array(
					'The system has detected that bttracker is NOT INSTALLED.',
					'The root of the bttracker must be placed in &lt;site root&gt;/include/bttracker/',
					'Bttracker is an open source torrent tracker.',
				),
			),
			'value' => array(
				'link' => array(
					'url' => 'http://phpbttrkplus.sourceforge.net/',
					'text' => 'Get PHP BTTracker+',
				),
			),
		);
	}
	
	return $status;
}

/**
 * Implementation of configure
 * @ingroup configure
 */
function configure_bt($settings)
{
	$settings['seeder_path'] = setting('seeder_path');
	
	$options = array();
	
	if(dependency('seeder'))
	{
		$options['seeder_path'] = array(
			'name' => lang('seeder title', 'Seeder'),
			'status' => '',
			'description' => array(
				'list' => array(
					lang('seeder description', 'This script requires that a seeder be installed.'),
				),
			),
			'type' => 'text',
			'value' => $settings['seeder_path'],
		);
	}
	else
	{
		$options['seeder_path'] = array(
			'name' => lang('seeder title', 'Seeder'),
			'status' => 'fail',
			'description' => array(
				'list' => array(
					lang('seeder description fail', 'The seeder specified is not installed or cannot be run.'),
				),
			),
			'type' => 'text',
			'value' => $settings['seeder_path'],
		);
	}
	
	return $options;
}

/**
 * Implementation of output
 * @ingroup output
 */
function output_bt($request)
{
	$request['bt'] = validate($request, 'bt');
	if($request['bt'] == 'announce')
	{
		global $admin_user, $admin_pass, $dbhost, $dbuser, $dbpass, $database;
		
		$_REQUEST = $request['bt_request'];
		
		// get administrator
		$admin = $GLOBALS['database']->query(array(
				'SELECT' => 'users',
				'WHERE' => 'id = -1',
				'LIMIT' => 1
			)
		, false);
	
		include_once setting('local_root') . 'include' . DIRECTORY_SEPARATOR . 'bttracker' . DIRECTORY_SEPARATOR . 'config.php';
		$admin_user = $admin[0]['Username'];
		$admin_pass = $admin[0]['Password'];
		
		$parsed_dsn = parseDSN(DB_CONNECT);
		$dbhost = $parsed_dsn['hostspec'];
		$dbuser = $parsed_dsn['username'];
		$dbpass = $parsed_dsn['password'];
		$database = $parsed_dsn['database'];
	
		include setting('local_root') . 'include' . DIRECTORY_SEPARATOR . 'bttracker' . DIRECTORY_SEPARATOR . 'announce.php';
		
	}
	else
	{
		include_once setting('local_root') . 'include' . DIRECTORY_SEPARATOR . 'bttracker' . DIRECTORY_SEPARATOR . 'BEncode.php';
		include_once setting('local_root') . 'include' . DIRECTORY_SEPARATOR . 'bttracker' . DIRECTORY_SEPARATOR . 'funcsv2.php';
	
		$request['cat'] = validate($request, 'cat');
		
		// make select call
		$files = get_files($request, $count, $request['cat']);
		
		$files_length = count($files);
		
		// the ids handler will do the replacement of the ids
		if(count($files) > 0)
			$files = get_ids(array('cat' => $request['cat']), $tmp_count, $files);
		
		// get all the other information from other handlers
		for($index = 0; $index < $files_length; $index++)
		{
			$file = $files[$index];
			$tmp_request = array();
			$tmp_request['file'] = $file['Filepath'];
		
			// merge with tmp_request to look up more information
			$tmp_request = array_merge(array_intersect_key($file, getIDKeys()), $tmp_request);
				
			// merge all the other information to each file
			foreach(get_handlers() as $i => $handler)
			{
				if($handler != $request['cat'] && handles($file['Filepath'], $handler))
				{
					$return = get_files($tmp_request, $tmp_count, $handler);
					if(isset($return[0])) $files[$index] = array_merge($return[0], $files[$index]);
				}
			}
		
			// for the second pass go through and get all the files inside of folders
			if(isset($file['Filetype']) && $file['Filetype'] == 'FOLDER')
			{
				// get all files in directory
				$props = array('dir' => $file['Filepath']);
				$sub_files = get_files($props, $tmp_count, $request['cat']);
				
				// put these files on the end of the array so they also get processed
				$files = array_merge($files, $sub_files);
				$files_length = count($files);
			}
		}
		
		// remove folders so we don't have to worry about them in the series of loops below
		foreach($files as $i => $file)
			if(isset($file['Filetype']) && $file['Filetype'] == 'FOLDER')
				unset($files[$i]);
		$files = array_values($files);
	
		
		if(count($files) > 0)
		{
			
			$total_size = 0;
			
			$torrent = array();
			$torrent['info'] = array();
			$torrent['info']['files'] = array();
			foreach($files as $index => &$file)
			{
				if(setting('admin_alias_enable') == true) $files[$index]['Filepath'] = preg_replace($GLOBALS['alias_regexp'], $GLOBALS['paths'], $file['Filepath']);
				if(!file_exists($files[$index]['Filepath']))
					continue;
				$file_info = array();
				$file_info['length'] = intval($file['Filesize']);
				$file_info['path'] = split('/', substr($file['Filepath'], strpos($file['Filepath'], '/') + 1));
				$file_into['md5sum'] = md5_file($file['Filepath']);
				$torrent['info']['files'][] = $file_info;
				$total_size += $file['Filesize'];
			}
			$torrent['info']['private'] = 0;
			// use a . as the name so that it can use the root of the file system
			$torrent['info']['name'] = 'Files from ' . setting('html_name');
			$torrent['info']['piece length'] = ($total_size < 1024*1024*10)?pow(2,18):(($total_size < 1024*1024*1024*10)?pow(2,19):pow(2,20));
			$torrent['info']['pieces'] = '';
			
			$current_file = 0;
			$contents = '';
			$read_amount = 0;
			$fp = output_handler($files[$current_file]['Filepath'], $request['cat']);
			while($current_file <= count($files))
			{
				$read_amount = $torrent['info']['piece length'] - strlen($contents);
			
				$contents .= fread($fp, $read_amount);
					
				//print $read_amount . '-' . strlen($contents) . '<br />';
				
				if(feof($fp))
					$current_file++;
				
				// append next file if centents are short
				if(feof($fp) && $current_file < count($files))
				{
					//print $files[$current_file]['Filepath'] . '<br />';
					$fp = output_handler($files[$current_file]['Filepath'], $request['cat']);
				}
				
				// should keep reading files if middle file is shorter then piece length also
				if(strlen($contents) == $torrent['info']['piece length'])
				{
					//print 'new<br/>';
					// sha and append to pieces
					$torrent['info']['pieces'] .= sha1($contents);
					$contents = '';
				}
			}
			if($contents != '')
				$torrent['info']['pieces'] .= sha1($contents);
			
			//print ceil($total_size / $torrent['info']['piece length']) * 20;
			
			$output = '';
			for($i = 0; $i < strlen($torrent['info']['pieces']); $i+=2)
			{
				$output .= chr(hexdec(substr($torrent['info']['pieces'], $i, 2)));
			}
			$torrent['info']['pieces'] = $output;
			
			$torrent['announce'] = url('bt/announce', true, true);
			$torrent['creation date'] = time();
			$torrent['comment'] = setting('html_name');
			$torrent['created by'] = setting('html_name');
			
			// encode info to sha1 it and get hash value
			$info_hash = sha1(BEncode($torrent['info']));
			
			// collect tracker info
			$properties = array(
				'info_hash' => $info_hash,
				'filename' => $torrent['info']['name']
			);
			$properties['info'] = $torrent['info']['piece length'] / 1024 * (strlen($torrent['info']['pieces']) / 20) /1024;
			$properties['info'] = round($properties['info'], 2) . " MB";
			if (isset($torrent['comment']))
				$properties['info'] .= " - " . $torrent["comment"];
		
			// insert torrent into tracker
			$query = 'INSERT INTO namemap (info_hash, filename, url, info) VALUES ("' . $properties['info_hash'] . '", "' . $properties['filename'] . '", "", "' . $properties['info'] . '")';
			makeTorrent($info_hash, true);
			quickQuery($query);
			
			// kill previous ctorrent
			$stop = "kill `cat ". escapeshellarg(setting('tmp_dir') . $info_hash . '.pid') . "`";
			system ($stop);
			if(file_exists(setting('tmp_dir') . $info_hash . '.pid'))
				unlink(setting('tmp_dir') . $info_hash . '.pid');
		
			
			// restart ctorrent
			$filename = 'temp-' . time();
			
			if($fp = @fopen(setting('tmp_dir') . $filename . '.torrent', 'w'))
			{
				fwrite($fp, BEncode($torrent));
				fclose($fp);
			}
			$pid = escapeshellarg(setting('tmp_dir') . $info_hash . '.pid');
			$file = escapeshellarg(setting('tmp_dir') . $filename . '.torrent');
			$stat = escapeshellarg(setting('tmp_dir') . $filename . '.stat');
			$command = "cd /tmp/; /usr/local/bin/ctorrent -s / '" . $file . "' > '" . $stat . "' & echo $! > '" . $pid . "'";
			shell_exec($command);
			
			header('Content-Type: application/x-bittorrent');
			header('Content-Disposition: filename="' . setting('html_name') . '-' . time() . '.torrent"'); 
			
			print BEncode($torrent);
		}
	}
}

